docs/tools/build_rules.md

```diff
@@ -0,0 +1,134 @@
+
+### Source Files
+
+Mbed OS Build tools determine what sort of source file it has found based on the file's extension. The following table lists all files that are used in a build.
+
+| Extension           | File Type         | Used by                        |
+|---------------------|-------------------|--------------------------------|
+| `.a`, `.ar`         | Archive           | Toolchain's Linker             |
+| `.c`                | C Source          | Toolchain's C Compiler         |
+| `.cc`, `.cpp`       | C++ Source        | Toolchain's C++ Compiler       |
+| `.h`, `.hpp`, `.hh` | C/C++ Header      | Toolchain's C and C++ Compiler |
+| `.icf`              | IAR Linker File   | IAR's Linker                   |
+| `.ld`               | GCC Linker Script | GCC's Linker (`ld`)            |
+| `.o`                | Object            | Toolchain's Linker             |
+| `.s`, `.S`          | Assembly          | Toolchain's assebler           |
+| `.sct`              | ARM Scatter File  | ARM's Linker (`armlink`)       |
+
+
+### Label Directories
+
+Label directories are directories that follow a naming convention: an upper case label type followed by an underscore followed by a label. The Mbed OS Build tools define 4 label types:
+
+ * `TARGET`, constructed from the configuration value `target.extra_labels` and the name of the targets.
+ * `FEATURE`, constructed from the configuration value `target.features`.
+ * `COMPONENT`, controlled from the configuration value `target.components`.
+ * `TOOLCHAIN`, controlled completely by the toolchain used to build.
+
+
+#### Target Directories
+
+The target labeled directories are used for selecting sources that are specific to an MCU or MCU family. The set of directories included in a given build is all of the names of the targets in the inheritance heirachy and the `target.extra_labels`. The following is an example `targets.json`:
+
+```json
+{
+    "MCUXPRESSO": {
+        "inherits": ["Target"],
+        "public": false
+    },
+    "TEENSY3_1": {
+        "inherits": ["MCUXPRESSO"],
+        "extra_labels": ["K20XX", "K20DX256"]
+    }
+}
+```
+
+The above example will include source files in directories named `TARGET_MCUXPRESSO` and `TARGET_TEENSY3_1`, from the inheritance heirachy and `TARGET_K20XX` and `TARGET_K20DX256` from the `target.extra_labels` value. It will not include source files within `TARGET_NORDIC`, `TARGET_K66F` or `TARGET_NUCLEO_F411` directories as they are not in the list of labels included in the label type `TARGET`
+
+#### Feature Directories
+
+The feature labeled directories are used for software that implements some funtionality that requires some amount of space even when it's not used. The set of directories include by the `FEATURE` label type is entirely controlled by the configuration value `target.features`. The following is a trimmed down version of an example `targets.json`:
```

Comment:
"The set of directories include by the `FEATURE`:" -> "The set of directories **included** by the `FEATURE`"